src/layers/core/path-layer/path-layer-vertex.glsl

```diff
@@ -1,71 +1,197 @@
 #define SHADER_NAME path-layer-vertex-shader
 
-attribute float directions;
 attribute vec3 positions;
-attribute vec3 leftDeltas;
-attribute vec3 rightDeltas;
 
-attribute vec4 colors;
-attribute vec3 pickingColors;
+attribute vec3 instanceStartPositions;
```

Comment:
We've discussed this off line, so most of this is clear now, but will still post my initial write-up since I wrote it.

I'm not reviewing this vertex shader in detail now, but as a general question, now that we use GL.TRIANGLES and pre-tesselate the line, is my assumption that this code is mainly there support rounded joins correct? 

On the perf side, I note that this shader contains a couple of if statements which is generally considered detrimental to GPU performance, especially when those if statements take different paths for different vertices which they clearly do here.

As an idea, it might be valuable to have @shaojingli review this shader, and also to instrument it with his GPU timer support and then experiment with various changes to understand how performance is affected.

Those changes could e.g. involve reorganizing the code or moving some calculations to JS in the tesselation stage to avoid branching here in GLSL, and then measuring again.